confignode/src/main/java/org/apache/iotdb/confignode/manager/DataNodeRemoveManager.java

```diff
@@ -0,0 +1,793 @@
+                } finally {
+                  if (req != null) {
+                    unRegisterRequest(req);
+                  }
+                }
+              }
+            });
+    workThread.setName("Exec-RemoveDataNode-Thread");
+  }
+
+  private void waitUntilBeLeader() {
+    if (waitLeaderThread != null) {
+      waitLeaderThread.start();
+    }
+  }
+
+  private void execRequestFormQueue() {
+    if (workThread != null) {
+      workThread.start();
+    }
+  }
+
+  public void beLeader() {
+    // TODO
+    /** this.isLeader = true; synchronized (leaderLock) { leaderLock.notify(); } */
+  }
+
+  public void beFollower() {
+    // TODO
+    /** this.isLeader = false; */
+  }
+
+  /**
+   * prepare for the request. from first node, first region to exec
+   *
+   * @param req RemoveDataNodeReq
+   */
+  private void prepareHeadRequestInfo(RemoveDataNodePlan req) {
+    LOGGER.info("start to prepare for request: {}", req);
+    this.headRequest = req;
+    // to exec the request, will change it's state at different stage.
+    this.headNodeIndex = 0;
+    this.headNodeState = DataNodeRemoveState.NORMAL;
+
+    TDataNodeLocation headNode = req.getDataNodeLocations().get(headNodeIndex);
+    this.headNodeRegionIds =
+        configManager.getPartitionManager().getAllReplicaSets().stream()
+            .filter(rg -> rg.getDataNodeLocations().contains(headNode))
+            .filter(rg -> rg.regionId.getType() != TConsensusGroupType.PartitionRegion)
+            .map(TRegionReplicaSet::getRegionId)
+            .collect(Collectors.toList());
+    this.headRegionIndex = 0;
+    this.headRegionState = RegionMigrateState.ONLINE;
+
+    // modify head quest
+    this.headRequest.setUpdate(true);
+    this.headRequest.setExecDataNodeRegionIds(headNodeRegionIds);
+    this.headRequest.setExecDataNodeIndex(headNodeIndex);
+    this.headRequest.setExecRegionIndex(headRegionIndex);
+    this.headRequest.setExecDataNodeState(headNodeState);
+    this.headRequest.setExecRegionState(headRegionState);
+    configManager.getConsensusManager().write(headRequest);
+    LOGGER.info("finished to prepare for request: {}", req);
+  }
+
+  private void setUp() {
+    removeQueue.addAll(configManager.getNodeManager().getDataNodeRemoveRequestQueue());
+    headRequest = configManager.getNodeManager().getHeadRequestForDataNodeRemove();
+    if (headRequest == null) {
+      return;
+    }
+
+    // avoid duplication
+    if (removeQueue.contains(headRequest)) {
+      removeQueue.remove(headRequest);
+    }
+
+    headNodeIndex = headRequest.getExecDataNodeIndex();
+    headNodeState = headRequest.getExecDataNodeState();
+    headNodeRegionIds = headRequest.getExecDataNodeRegionIds();
+    headRegionIndex = headRequest.getExecRegionIndex();
+    headRegionState = headRequest.getExecRegionState();
+  }
+
+  /**
+   * exec the request loop all removed node 1: brocast it to cluster 2: loop region on it 2.1 change
```

Comment:
The code doesn't have 2.1 change region leader. we can delete comment 2.1 so it corresponds to code behavior.

Reply:
fixed